File: js/sandbox/control.js
/**
 * @file JSLAB library control submodule
 * @author Milos Petrasinovic <mpetrasinovic@pr-dc.com>
 * PR-DC, Republic of Serbia
 * info@pr-dc.com
 */
 
/**
 * Class for JSLAB control submodule.
 */
class PRDC_JSLAB_LIB_CONTROL {
  
  /**
   * Initializes the control submodule.
   * @param {Object} jsl Reference to the main JSLAB object.
   */
  constructor(jsl) {
    var obj = this;
    this.jsl = jsl;
  }

  /**
   * Create a transfer function representation.
   * @param {number[]} num - Numerator coefficients of the transfer function.
   * @param {number[]} den - Denominator coefficients of the transfer function.
   * @param {number} [Ts=0] - Sampling time, defaults to 0 for continuous-time systems.
   * @returns {object} An object representing the transfer function { num, den, Ts }.
   */
  tf(num, den, Ts = 0) {
    return { num, den, Ts };
  }

  /**
   * Create a state-space representation.
   * @param {number[][]} A - System matrix.
   * @param {number[][]} B - Input matrix.
   * @param {number[][]} C - Output matrix.
   * @param {number[][]} D - Feedthrough matrix.
   * @param {number} [Ts=0] - Sampling time, defaults to 0 for continuous-time systems.
   * @returns {object} An object representing the state-space system { A, B, C, D, Ts }.
   */
  ss(A, B, C, D, Ts = 0) {
    return { A, B, C, D, Ts };
  }

  /**
   * Convert a transfer function to a state-space representation.
   * @param {number[]} num - Numerator coefficients of the transfer function.
   * @param {number[]} den - Denominator coefficients of the transfer function.
   * @returns {object} State-space representation of the system.
   */
  tf2ss(num, den) {
    const n = den.length;
    num = [...zeros(n - num.length), ...num];

    let A = zeros(n-1).map(() => zeros(n-1));
    for(let i = 1; i < n - 1; i++) A[i-1][i] = 1;
    A[n - 2] = fliplr(den.slice(1).map(d => -d));

    const C = fliplr(num.map((c, i) => c - den[i] * num[0]).slice(1));
    const B = zeros(n - 2).concat([1]);
    const D = num[0];

    return this.ss(A, B, C, D);
  }

  /**
   * Convert a state-space representation to a transfer function.
   * @param {object} sys - State-space system { A, B, C, D }.
   * @returns {object} Transfer function representation { num, den }.
   */
  ss2tf(sys) {
    const { A, B, C, D } = sys;
  
    const den = charpoly(A);
    const num = plus(charpoly(minus(A, this.jsl.env.math.multiply(B, [C]))),
      scale(den, D-1));

    var p = den.length - num.length;
    if(p > 0) {
      num = [...zeros(p), ...num];
    }
    return { num, den };
  }

  /**
   * Convert a continuous-time transfer function to discrete-time.
   * @param {number[]} numc - Continuous-time numerator coefficients.
   * @param {number[]} denc - Continuous-time denominator coefficients.
   * @param {number} Ts - Sampling time.
   * @returns {object} Discrete-time transfer function representation { num, den }.
   */
  c2d(numc, denc, Ts) {
    const sysc = this.tf2ss(numc, denc);
    const sysd = this._c2dZOH(sysc, Ts);
    return this.ss2tf(sysd);
  }

  /**
   * Convert a continuous-time state-space system to discrete-time using Zero-Order Hold.
   * @param {object} sysc - Continuous-time state-space system { A, B, C, D }.
   * @param {number} Ts - Sampling time.
   * @returns {object} Discrete-time state-space system { A, B, C, D, Ts }.
   */
  _c2dZOH(sysc, Ts) {
    const { A, B, C, D } = sysc;
    const n = A.length;

    const M = A.map((row, i) => [...row, B[i]]);
    M.push(zeros(n + 1));
    const M_exp = jsl.env.math.expm(jsl.env.math.multiply(M, Ts))._data.slice(0, n);

    const Ad = M_exp.map(row => row.slice(0, n));
    const Bd = M_exp.map(row => row.slice(n));

    return this.ss(Ad, Bd, C, D, Ts);
  }
  
  /**
   * Simulate the time response of a discrete-time linear system.
   * @param {number[]} sys - transfer function of system.
   * @param {number[]} u - Input signal array.
   * @param {number[]} t - Time vector array.
   * @returns {object} An object containing the response:
   *                   - y: Output signal array.
   *                   - t: Time vector array (same as input).
   */
  lsim(sys, u, t, Ts) {
    var sysd = sys;
    if(!sys.Ts && Ts) {
      sysd = this.c2d(sys.num, sys.den, Ts);
    }
    var num = sysd.num;
    var den = sysd.den;

    // Normalize the denominator coefficients so that den[0] = 1
    if(den[0] !== 1) {
      const den0 = den[0];
      for(let i = 0; i < den.length; i++) {
        den[i] /= den0;
      }
      for(let i = 0; i < num.length; i++) {
        num[i] /= den0;
      }
    }

    const N = u.length;
    var y = zeros(N); // Initialize output array with zeros

    // Iterate over each time step starting from n=1
    for(let n = 0; n < N; n++) {
      // Compute feedforward terms: sum(num[k] * u[n - k}) for k =0 to M
      for(let k = 0; k < num.length; k++) {
        const idx = n - k;
        if(idx >= 0) {
          y[n] += num[k] * u[idx];
        }
      }

      // Compute feedback terms: sum(den[k} * y[n - k}) for k =1 to N
      for(let k = 1; k < den.length; k++) {
        const idx = n - k;
        if(idx >= 0) {
          y[n] -= den[k] * y[idx];
        }
      }
    }
    
    if(num.length < den.length) {
      var n = den.length - num.length;
      y = [...zeros(n), ...y.slice(0, y.length - n)];
    }
    
    return { y, t };
  }
  
  /**
   * Simulates the system response over a specified time period.
   * @param {Object} sys - The system to simulate.
   * @param {number} Tfinal - The final time for the simulation.
   * @returns {Object} The simulation result.
   */
  step(sys, Tfinal) {
    var u = ones(101);
    var t = linspace(0, Tfinal, 101);
    var Ts = Tfinal/100;
    return this.lsim(sys, u, t, Ts);
  }
  
  /**
   * Estimates a transfer function model using numerical optimization.
   * @param {Array<number>} t - Time vector.
   * @param {Array<number>} u - Input signal vector.
   * @param {Array<number>} y - Output signal vector.
   * @param {number} np - Number of poles.
   * @param {number} nz - Number of zeros.
   * @param {string} [method='NelderMead'] - Optimization method ('NelderMead' or 'Powell').
   * @returns {{sys: object, error: number}} Estimated transfer function and mean squared error.
   */
  tfest(t, u, y, np, nz, method = 'NelderMead') {
    var obj = this;
    var Ts = mean(diff(t));
    var N = np+nz+1;
    var f = (x) => { // funkcija prilagodjenosti
      try {
        var den = [1, ...x.slice(0, np)];
        var num = x.slice(np);
        var sys = tf(num, den);
        var r = obj.lsim(sys, u, t, Ts);
        var mse = obj.jsl.math.mse(y, r.y);
        return mse;
      } catch(err) {
        return Infinity;
      }
    };
    var r;
    if(method == 'NelderMead') {
      r = this.jsl.optim.optimNelderMead(f, zeros(N));
    } else if(method == 'Powell') {
      r = this.jsl.optim.optimPowell(f, zeros(N));
    }
    var den = [1, ...r.x.slice(0, np)];
    var num = r.x.slice(np);
    return {sys: tf(num, den), error: r.fx};
  }
}

exports.PRDC_JSLAB_LIB_CONTROL = PRDC_JSLAB_LIB_CONTROL;

File: dev/control.m
clc, clear all, close all
format long
pkg load control
pkg load signal

function [y, t_out] = lsim_custom(num, den, u, t_in)
  % lsim_custom - Custom implementation of lsim for discrete-time systems
  %
  % Syntax:
  %   [y, t_out] = lsim_custom(num, a, u, t_in)
  %
  % Inputs:
  %   num     - Numerator coefficients of the transfer function (vector)
  %   den     - Denominator coefficients of the transfer function (vector)
  %   u     - Input signal (vector)
  %   t_in  - Time vector (vector)
  %
  % Outputs:
  %   y     - Output signal (vector)
  %   t_out - Time vector (same as input)

  % Number of samples
  N = length(u);

  % Initialize output vector with zeros
  y = zeros(N, 1);

  % Ensure denominator den[1] is 1
  if den(1) ~= 1
    num = num / a(1);
    den = den / den(1);
  end

  % Compute the output using the difference equation
  for n = 1:N
    % Feedforward (num) terms: sum num(k) * u[n - k + 1]
    for k = 1:length(num)
      idx = n - k + 1;
      if idx >= 1
        y(n) += num(k) * u(idx);
      end
    end

    % Feedback (den) terms: sum den(k) * y[n - k + 1] for k >= 2
    for k = 2:length(den)
      idx = n - k + 1;
      if idx >= 1
        y(n) -= den(k) * y(idx);
      end
    end
  end

  if(length(num) < length(den))
    n = length(den) - length(num)
    y = [zeros(n, 1); y(1:end-n)];
  end

  % Output time vector remains unchanged
  t_out = t_in;
end

function sys = tf_custom(num, den, Ts)
    % tf_custom Creates a transfer function structure
    %
    % Syntax:
    %   sys = tf_custom(num, den, Ts)
    %
    % Inputs:
    %   num - Numerator coefficients
    %   den - Denominator coefficients
    %   Ts  - Sampling time (optional, default is 0 for continuous-time)
    %
    % Output:
    %   sys - Transfer function structure with fields num, den, Ts

    if nargin < 3
        Ts = 0; % Continuous-time by default
    end

    sys.num = num;
    sys.den = den;
    sys.Ts = Ts;
end

function sys = ss_custom(A, B, C, D, Ts)
    % ss_custom Creates a state-space structure
    %
    % Syntax:
    %   sys = ss_custom(A, B, C, D, Ts)
    %
    % Inputs:
    %   A  - State matrix
    %   B  - Input matrix
    %   C  - Output matrix
    %   D  - Feedthrough matrix
    %   Ts - Sampling time (optional, default is 0 for continuous-time)
    %
    % Output:
    %   sys - State-space structure with fields A, B, C, D, Ts

    if nargin < 5
        Ts = 0; % Continuous-time by default
    end

    sys.A = A;
    sys.B = B;
    sys.C = C;
    sys.D = D;
    sys.Ts = Ts;
end

function sys = tf2ss_custom(num, den)
  n = length(den);
  num = [zeros(1, length(den) - length(num)), num];

  A = eye(n-1);
  A(1, :) = [];
  A(n-1, :) = -fliplr(den(2:(n)));

  C = num-den.*num(1);
  C = fliplr(C(2:(n)));

  B = zeros(n-1, 1);
  B(end) = 1;

  D = num(1);

  sys = ss_custom(A, B, C, D);
end

function [num, den] = ss2tf_custom(sys)
    A = sys.A;
    B = sys.B;
    C = sys.C;
    D = sys.D;
    Ts = sys.Ts;

    % Calculate the denominator
    den = poly(A);

    % Calculate the numerator
    num = poly(A - B * C) + (D-1)*den;

    % Normalize numerator length
    num = [zeros(1, length(den) - length(num)), num];
end

function sysd = c2d_zoh_custom(sysc, Ts)
    % c2d_zoh_custom Discretizes a continuous-time state-space system using Zero-Order Hold (ZOH)
    %
    % Syntax:
    %   sysd = c2d_zoh_custom(sysc, Ts)
    %
    % Inputs:
    %   sysc - Continuous-time state-space structure with fields A, B, C, D
    %   Ts   - Sampling time
    %
    % Output:
    %   sysd - Discrete-time state-space structure with fields A, B, C, D, Ts

    A = sysc.A;
    B = sysc.B;
    C = sysc.C;
    D = sysc.D;

    n = size(A,1);
    m = size(B,2);

    % Construct the augmented matrix
    M = [A, B; zeros(m, n + m)];

    % Compute the matrix exponential
    M_exp = expm(M * Ts);

    Ad = M_exp(1:n, 1:n);
    Bd = M_exp(1:n, n+1:end);

    % Update D to account for the sampling interval
    Dd = D;

    % Create the discrete-time system
    sysd = ss_custom(Ad, Bd, C, Dd, Ts);
end

function [numd, denda] = c2d_custom(numc, denc, Ts)
    % c2d_custom Discretizes a continuous-time transfer function using ZOH
    %
    % Syntax:
    %   [numd, denda] = c2d_custom(numc, denc, Ts)
    %
    % Inputs:
    %   numc - Numerator coefficients of the continuous-time transfer function
    %   denc - Denominator coefficients of the continuous-time transfer function
    %   Ts   - Sampling time
    %
    % Outputs:
    %   numd  - Numerator coefficients of the discrete-time transfer function
    %   denda - Denominator coefficients of the discrete-time transfer function

    % Convert transfer function to state-space using custom tf2ss
    sysc = tf2ss_custom(numc, denc);

    % Discretize the state-space system using custom ZOH
    sysd = c2d_zoh_custom(sysc, Ts);

    % Convert discrete state-space to transfer function using custom ss2tf
    [numd, denda] = ss2tf_custom(sysd);
end

% System transfer function
num = [3.953, 1];        % Numerator
den = [1, 2, 3.868];  % Denominator

% Time vector and input signal
t = 0:49;       % 50 time steps
u = ones(size(t)); % Unit step input

% Response using built-in lsim
sys = tf(num, den, 1); % Transfer function with sampling time of 1
[y_builtin, t_builtin, x] = lsim(sys, u, t);

% Response using custom lsim
[y_custom, t_custom] = lsim_custom(num, den, u, t);

[y_builtin, y_custom]

num = [1.095, 23.03];        % Numerator
den = [1, 16.52, 155.1, 2];  % Denominator
sys1 = tf(num, den); % Transfer function
c2d(sys1, 1)
[numd, dend] = c2d_custom(num, den, 1)

num = [1.095, 23.03, 3];        % Numerator
den = [1, 16.52, 155.1, 2];  % Denominator
sys1 = tf(num, den); % Transfer function
c2d(sys1, 1)
[numd, dend] = c2d_custom(num, den, 1)

num = [1.095, 23.03, 1, 1];        % Numerator
den = [1, 16.52, 155.1, 2];  % Denominator
sys1 = tf(num, den); % Transfer function
c2d(sys1, 1)
[numd, dend] = c2d_custom(num, den, 1)

A = [-7 1 0; -14 0 1; -8 0 0];
B = [20; 125; 185];
C = [1 0 0 ];
D = 5;

sys = ss(A, B, C, D);
disp('SYS1');
[num, den] = ss2tf(sys)
[num, den] = ss2tf_custom(sys)

A = [0 1 0; 0 0 1; -8 -14 -7];
B = [0; 0; 1];
C = [15 5 0];
D = 0;

sys = ss(A, B, C, D);
disp('SYS2');
[num, den] = ss2tf(sys)
[num, den] = ss2tf_custom(sys)

disp('c2d');
num = [3.8978];        % Numerator
den = [1, 3.8140];  % Denominator
c2d(tf(num, den), 1)

num = [1.0193, 22.4826, 547.6238];        % Numerator
den = [1, 16.3824, 152.4330, 548.0400];  % Denominator
c2d(tf(num, den), 1)

